# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

This supervised learning problem is one of **classification**, since the desired output is whether (in a binary sense) a student needs early intervention.  
This supervised learning problem does not seek to quantify or convey the strength of the need for intervention, which would be a regression, but instead merely flags (or fails to flag) a student as needing early intervention, under the presumption that the school's administration then intervenes commensurate with the student's need.  
In machine learning parlance, the desired output here is a **label**, either yes or no, on the **class** of "Does this student need early intervention"?

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [2]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


In [3]:
student_data.head()
# Get an idea of the shape of the dataset

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   ...   internet romantic  famrel  freetime  goout Dalc Walc health absences  \
0  ...         no       no       4         3      4    1    1      3        6   
1  ...        yes       no       5         3      3    1    1      3        4   
2  ...        yes       no       4         3      2    2    3      3       10   
3  ...        yes      yes       3         2      2    1    1      5        2   
4  ...         no       no       4         3      2    1    2      5        4   

  passed  
0     no  
1     no  
2    yes  
3    yes  
4    yes  

[5 rows x 31 columns]

In [4]:
student_data.describe()
# Another view of the data, wholly for my own personal benefit

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences  
count  395.000000  
mean     5.708861  
std      8.003096  
min      0.000000  
25%      0.000000  
50%      4.000000  
75%      8.000000  
max     75.000000

In [5]:
student_data.isnull().apply(sum)
# Check whether the dataset is missing information

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
passed        0
dtype: int64

### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [6]:
# TODO: Calculate number of students
n_students = len(student_data)

# TODO: Calculate number of features
n_features = len(student_data.columns) - 1
    # We are told in the prompt that one and only one column is a label,
    # hence this must be subtracted from the count.

# TODO: Calculate passing students
n_passed = sum(student_data["passed"]=="yes")

# TODO: Calculate failing students
n_failed = sum(student_data["passed"]=="no")

# TODO: Calculate graduation rate
grad_rate = float(n_passed) / n_students * 100
    # Python 2.7 defaults to integer division.

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [7]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [8]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [9]:
# TODO: Import any additional functionality you may need here

# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
from sklearn.cross_validation import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X_all, y_all, stratify=y_all, train_size = num_train, test_size = num_test, random_state = 42)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

# Verify that the stratification was successful, directly pulled from review
print ""
print "Grad rate of the training set: {:.2f}%".format(100 * (y_train == 'yes').mean())
print "Grad rate of the testing set: {:.2f}%".format(100 * (y_test == 'yes').mean())

Training set has 300 samples.
Testing set has 95 samples.

Grad rate of the training set: 67.00%
Grad rate of the testing set: 67.37%


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

Three supervised learning models that are appropriate for this problem are **Decision Trees**, **Gradient Boosting**, and **K-Nearest Neighbors**, discussed below.
<h3><font color="darkgreen">Decision Trees</font></h3>
**Decision Trees** attempt to classify the data by playing 20 questions.  This algorithm works best when there are easily-identified clefts along a feature, and the features are fairly independent of each other.  This allows each question to generate maximal information gain, and reduces having to ask the same or a similar question again later.
<h4>Uses</h4>
Probably the most impressive collection of scientific and industrial applications of Decision Trees is found in [this 1995 write-up](http://www.cbcb.umd.edu/~salzberg/docs/murthy_thesis/survey/node32.html) by big-data rockstar Sreerama K. Murthy, which includes applications in the fields of astronomy, manufacturing, pharmacology, and medicine.  One particular such application was to [classify objects](http://iopscience.iop.org/article/10.1086/133551/meta) in Hubble Space Telescope images as "stars, galaxies, cosmic rays, plate defects, and other types of objects ... with over 95% accuracy using data from a single, upaired image".
<h4>Strengths</h4>
Decision Trees are conceptually appealing; if we have a binary classification for which there are a host of necessary and sufficient conditions, we can concisely represent this classification exercise as a decision tree.  
Similarly, we can represent arbitrary complexity with decision trees (although we may not always want to do so).  Finally, it's possible to tune a Decision Tree to the "right" degree of complexity for the dataset.
<h4>Weaknesses</h4>
Fundamentally, a decision tree is not a good solution when the label boundaries are not feature-axis-aligned.  That is, if y = x is the best classification boundary, a decision tree would need many tiny "stairs" to capture that fact.  This would result in a perfectly-functional decision tree, but which requires traversing a computationally large number of nodes to arrive at the correct classification.  
Further, the strength of being able to tune a Decision Tree to the "right" degree of complexity is, in the same breath, a weakness that allows for a Decision Tree to be improperly tuned, either leading to overgeneralizations that miss most of the training data or to hyper-specializations that overfit at the expense of predictive power.  
<h4>Application to Student Intervention</h4>
Here, I would presume that certain features included with this dataset have meaningful classification clefts.  For example, if you've missed every day of school, I feel comfortable predicting that you need intervention to pass.  Similarly, if your weekday alcohol consumption is high, that probably means that intervention is needed to pass.  Other intuitive-feeling clefts would be low weekly study time without being a genious, very bad family relationships, and not wanting to take higher education.  
Now, the decision tree algorithm doesn't take _my_ input for what are useful features along which to cleave, but I do think those axis-aligned split points exist, and so I would guess that the decision tree algorithm is a good classifier to train on this dataset.

<h3><font color="darkgreen">Gradient Boosting</font></h3>
The **Gradient Boosting** algorithm tries to learn from what it gets wrong, by iteratively doing slightly better than random chance at guessing on the leftover error.  That doesn't sound terribly impressive, but in practice it has been quite useful.  
This algorithm is very similar to the AdaBoost algorithm discussed in lecture, with a technical difference in _how_ it focuses on leftover error (i.e. gradient boost employs gradient descent on the sum of squared residual errors, while AdaBoost tweaks the "importance" of datapoints based on the correctness of the previous classification). 
<h4>Uses</h4>
Gradient Boosting has been used successfully for neurorobotics applications, specifically in mapping [EMG](https://en.wikipedia.org/wiki/Electromyography) signals to [a robotic arm](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3885826/) by watching the human participant and measuring brain activity.  Gradient Boosting is also used extensively in facial recognition, whether in [realtime](http://accord-framework.net/docs/html/T_Accord_Vision_Detection_HaarObjectDetector.htm), or to catalog who was present in a [video-steam](http://link.springer.com/chapter/10.1007%2F978-3-642-23123-0_14).  Another visual application of the Gradient Boosting approach is to [detect crop blight](http://imperialjournals.com/index.php/IJIR/article/view/465/449).
<h4>Strengths</h4>
Being an ensemble method, Gradient Boosting is versatile in that it can use any of a wide variety of base classifiers, from the common decision tree to the inscrutible neural network.  
As suggested by the above applications, Gradient Boosting with decision trees has the advantage of being able to work well without the manual injection of domain knowledge.  The base classifier can of course still be poorly-suited to the classification task at hand, but gradient boosting typically makes the best of whatever classifier it's given.  
Gradient Boosting is also the algorithm [_du jour_](https://github.com/dmlc/xgboost/tree/master/demo#machine-learning-challenge-winning-solutions) for Kaggle.  That's not a rigorous examination of the merits of the algorithm, but it does speak to the machine learning community's adoption of it, and thus to the likelihood of (1) finding a colleague who understands the algorithm, of (2) finding a robust and well-tested software implementation of the algorithm, and of (3) future practical and theoretical work being done on this algorithm.
<h4>Weaknesses</h4>
Gradient Boosting is not magic; it is still subject to the constraints of garbage-in:garbage-out.  That is, it requires a weak classifier even to work, and this is not an easy requirement to meet.
The algorithm also [doesn't](http://www.cs.columbia.edu/~rocco/Public/mlj9.pdf) handle random noise well.  That is, boosting algorithms tend to kind of obsess over data that it's not correctly classifying, without any mechanism to ignore any one label as erroneous.  
This algorithm is also memory-intensive.  This is not surprising, since it has to remember at least some of perhaps thousands of individual weak classifiers to comprise the composite classifier that is the goal of boosting.  While it's a truism that memory is cheap, it's also the case that we will find increasingly larger problems to keep pace with the advances in memory capacity and access speed.
<h4>Application to Student Intervention</h4>
Here, the idea is that individual, say, Decision Tree classifiers, will be sufficiently good to be classified as "weak" on the student intervention dataset.  According to the algorithmic requirements of gradient boosting, that's all that's needed.  The assumption is of course that the Gradient Boosting algorithm will appropriately stitch together individual Decision Tree classifiers to minimize the sum of squared prediction errors, and in so doing, produce a composite classifier with a high degree of accuracy on the test set.

<h3><font color="darkgreen">K-Nearest Neighbors</font></h3>
**K-Nearest Neighbors** attempts to provide a classification label based off of some measure of **similarity**.  
That is, if something has feathers, quacks, and really loves bread, it's probably a duck, even if that's a bit of an oversimplification.
<h4>Uses</h4>
KNeighbors has been used successfully in [medical research](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2243774/) and [credit card fraud detection](https://www.researchgate.net/publication/236962626_credit_card_fraud_detection_using_anti_k-nearest_algorithm), but could reasonably be expected to produce good results whenever there is a diffusion of datapoints in a particular feature space (read: enough data), within some contiguous space-saving boundary (that is, not highly intermixed).
<h4>Strengths</h4>
Because K-Nearest Neighbors depends merely on some measure of similarity, it is possible for a KNeighbors classifier to accurately capture arbitrary structural complexity within the data, without the need to inject that domain knowledge in the form of a kernel trick, polynomial degree, or some other function.  
This allows a KNeighbors algorithm to merely "discover" structure in the data, once a measure of similarity is supplied.  
When datapoints which are similar in the featurespace are also similar in the labelspace, KNeighbors is an excellent algorithm.
<h4>Weaknesses</h4>
KNeighbors has difficulty with edge cases, such as whether someone with really high cholesterol who exercises regularly and eats well would be at risk for heart disease, or whether a particular entry in the [iris dataset](https://en.wikipedia.org/wiki/Iris_flower_data_set#/media/File:Iris_dataset_scatterplot.svg) with median sepal length and sepal width is Iris virginica or Iris versicolor, or whether someone who is socially conservative but supports broad spending on social programs is a "Republican" or "Democrat".  
Further, KNeighbors suffers from the **curse of dimensionality**, where the number of datapoints needed to properly train a KNeighbors algorithm grows exponentially with the number of features (dimensions) given to the classifier.  That is, KNeighbors is really only feasible with less than 10 input features, unless the number of input data is billions upon billions.  
Here, we have 300 training points.  That suggests we can probably only hope to use 5 features max, and those may need to be combined / projected featurespaces.
<h4>Application to Student Intervention</h4>
Here, the assumption is that there is some similarity measure between students who are numerically close to each other in some representation of a feature space.  
That's probably not as weird as it sounds for the clear-cut cases, but the concern of course is whether there is sufficient "distance" between passing and non-passing students to allow for proper classification.  If there's significant spatial overlap, it may be difficult for KNeighbors to correctly classify border cases.  This increases the percentage of students who would be susceptible to Type 1 neglect or Type 2 over-worry.  
In the alternative, it could be that there is a significant spatial boundary between students who require and students who do not require early intervention, in which case the KNeighbors algorithm should function exceptionally well on this dataset.
<br><br>
<font color="darkred">_Discarded candidates_</font>:
<font color="darkred">**Gaussian Naive Bayes**</font> is an attempt to provide classifications based on the "naive" assumption of conditional independence between the input features.  Here, that seems inappropriate, considering that many of the features are likely outgrowths of other, simpler features (e.g. famsize, address, Medu, Fedu, Mjob, Fjob, activities, nursery, internet, goout, health, etc. likely co-vary with parental income).

<font color="darkred">**Stochastic Gradient Descent**</font> models attempt to discover the set of parameters which minimize the error of the prediction.  However, they do not provide information on what the error "surface" looks like.  That is, (1) there is no guarantee that the minimum which is found is the global, rather than merely a local, minimum, (2) there is no guarantee that the minimum which is found is even "good" relative to other regional minimums, and (3) it could be that the error surface is like a rolling sand dune, where no particular minimum is good, or even much better than other possible minimums.  It is possible (and quite effective), through multiple randomized initiations, to merely take thousands of samples of the existing local minimums, and to compare them to each other, in an attempt to locate the global minimum, but this may be computationally very expensive.

<font color="darkred">**Support Vector Machines**</font> rely on linear or near-linear separability of the data (as represented in some-dimensional space).  If the data is not linearly separable in the default space, it requires a transformation via a kernel method into a space in which it is linearly separable.  The main problem with that fact is that here I don't know, _a priori_, how the features affect whether a student passes or doesn't.  So, if the default spatial representation of the data results in a poorly-performing SVM, I wouldn't actually know which kernel method to apply to increase performance, even acknowledging that there probably is one which would allow for excellent SVM performance.  Visualizing data in a three-dimensional feature space is hard enough, but here we start with a 30-dimensional feature space, which my tiny human brain struggles to comprehend visually.

<font color="darkred">**Logistic Regression**</font> (actually a classification) is a lot like linear regression, except that the continuous variable it outputs is the _likelihood_ of whether a datapoint does or doesn't (in a binary sense) have a label.  Here, that would mean that a properly-implemented Logistic Regression would output the likelihood that a student does or doesn't need early intervention.  The output would be useful, but the underlying assumption is that the variables in the feature space have a consistent linear relation to our output label, which for many of the labels seems like a ludicrous assumption.

### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [10]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [11]:
# TODO: Import the three supervised learning models from sklearn
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier

# TODO: Initialize the three models
clf_A = DecisionTreeClassifier(random_state=42)
clf_B = GradientBoostingClassifier(random_state=42)
clf_C = KNeighborsClassifier()
    # Presumes, without justification, NNeighbors = 5

# TODO: Set up the training set sizes
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train
y_train_300 = y_train

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)
for clf in [clf_A, clf_B, clf_C]:
    for xtrain, ytrain in [(X_train_100,y_train_100),
                           (X_train_200,y_train_200),
                           (X_train_300,y_train_300)]:
        train_predict(clf, xtrain, ytrain, X_test, y_test)
        print""

Training a DecisionTreeClassifier using a training set size of 100. . .
Trained model in 0.0056 seconds
Made predictions in 0.0005 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.6452.

Training a DecisionTreeClassifier using a training set size of 200. . .
Trained model in 0.0024 seconds
Made predictions in 0.0003 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.7258.

Training a DecisionTreeClassifier using a training set size of 300. . .
Trained model in 0.0034 seconds
Made predictions in 0.0004 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0003 seconds.
F1 score for test set: 0.6838.

Training a GradientBoostingClassifier using a training set size of 100. . .
Trained model in 0.0680 seconds
Made predictions in 0.0007 seconds.
F1 score for training set: 1.0000.
Made predictions in 0.0006 seconds.
F1 score for test set: 0.7500.

Training a GradientBoost

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - DecisionTreeClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0022           |     0.0006             |   1.0000         |    0.6452       |
| 200               |        0.0025           |     0.0006             |   1.0000         |    0.7258       |
| 300               |        0.0060           |     0.0005             |   1.0000         |    0.6838       |

** Classifer 2 - GradientBoostClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0785              |    0.0007              |    1.0000        |   0.7500        |
| 200               |     0.0893              |    0.0010              |    0.9924        |   0.7313        |
| 300               |     0.1195              |    0.0020              |    0.9781        |   0.7463        |

** Classifer 3 - KNeighborsClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0007              |    0.0020              |   0.8252         |  0.7586         |
| 200               |     0.0009              |    0.0051              |   0.8097         |  0.7857         |
| 300               |     0.0012              |    0.0106              |   0.8539         |  0.8138         |

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

The goal here is to find the best-performing model.

That goal is similar to finding the best-performing student:  We ask the model to complete a test, then we score it on how well it did.  The test is similar to, but not the same as, the study materials we gave to the model ahead of time, because we want the model to be able to generalize its lessons outside of specific problems we assigned it.

Score-wise, the "best" model in the preliminary assessment is the **KNeighborsClassifier**, which scores around 81% (way better than random guessing!).  However, **GradientBoostingClassifier** is close behind at ~75% on a preliminary assessment.  Given the number of hyperparameters left to tune away from the default settings, scores this close should be interpreted to signal a rough equivalency in performance capability, and likely **not** as a definitive answer as which model will perform best after tuning.

Other valid considerations are how expensive it is to "teach" the model, and how expensive it is for the model to provide us with a meaningful answer.  In this application, this model will likely be "taught" once, then asked for meaningful answers numerous times.  So here, we care way more about the expense of the answer than we do about the expense of the learning.

In this respect, the **GradientBoostingClassifier** is the clear winner: its prediction time (the expense of the answer) is ~1/5 the prediction time of the KNeighborsClassifier.  The training time (the "learning" expense) of the GradientBoostingClassifier is ~100x that of the KNeighborsClassifier, but that's a one-time occurrence and thus is far less of a concern than the expense of an answer.

Thus, **GradientBoostingClassifier** is presumed to be the best model, as its accuracy is roughly equivalent to the KNeighborsClassifier, and as its prediction time is a fraction of that algorithm.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

<h4>Motivation and Assumptions</h4>
We want all of the students to graduate.  A major obstacle to graduation is not receiving timely academic intervention.  Knowing earlier which students need intervention to pass is thus a top priority.

By applying machine learning techniques, we are able to accurately predict which students require academic intervention in order to pass, solely by examining information from the student's academic record supplemented with information provided by a student survey.

This prediction is possible because whether or not a student needs intervention is not random, but rather depends quantifiably on a variety of other factors such as parental involvement and education, extracurricular activities, school absences, etc.  By examining which students have failed in the past, this model is able to understand which students are currently at risk of failing.

At regular intervals, this model will examine each enrolled student for signs that the student may need intervention in order to pass.  The model will flag students that are at risk of failing without intervention, so that the administration can intervene before it is too late to help the student.

Exactly how this occurs is complicated, but clever.

<h4>A Guessing Game</h4>
We can make guesses about which factors determine whether a student needs intervention, and exactly how they combine to do so.

In fact, we can make our first guess _right now_ about how the different student features generally map to needing or not needing intervention: "Everyone who has failed at least one class before needs an intervention in order to pass, and everyone who hasn't failed a class before does not need an intervention in order to pass."  

Or, maybe it's that "Everyone who has 10 or more absences needs intervention in order to pass, and everyone with less than 10 absences doesn't need intervention in order to pass."

Or, maybe it's a combination of the above two, or more than that.

It's not clear at present, so let's talk about how to tell how well we're guessing.

<h4>Getting Good at the Guessing Game</h4>
We evaluate our guess by looking at how our predictions for who needed intervention compares to who actually needed intervention.  Specifically, for each student, we subtract what we should have predicted from what we did predict, then we square the difference.

Squaring the difference is a bit of cleverness, which allows us to know along which "[direction](https://www.youtube.com/watch?v=eikJboPQDT0)" to change our guess along in order to make less-bad predictions.  Specifically, this is accomplished through the computation of the gradient at that guess-point in the guess-space.  While the gradient points up-hill, here we're looking to minimize our squared error, so we just go the exact opposite direction of the gradient. 

Since we've figured out the direction to tweak our guessing so that we make better guesses, we then take what we hope is an appropriately-sized step in that direction.  It's possible we're stepping too large (in which case we might overshoot the best-guess-point we're aiming at), and it's also possible that we're stepping too small (in which case it's going to take us a loooong time to get where we're going, but we're definitely not going to miss it).

Eventually, we'll arrive at the best guess in that area.  

Perfect, right?  Well, not exactly; we don't know whether it's the best guess we could possibly make, and there's still some predictions that we're getting way wrong.  Is there anything we can do to "boost" our performance on the "residual" errors?

<h4>Boosting our Performance</h4>
Weirdly, yes there is, and it's called "boosting."  What we've done above is identified a guess that provides us with some accurate predictions about whether a student needs intervention, but there are some predictions it's still wrong about.  Let's call this leftover inaccuracy "remaining" error, and in fact, let's play our guessing game again, but this time just playing with the "residual" error.

Since we were able to generate accuracy that was better than totally random guessing with the student data in the first game, it must also be the case that we'll be able to generate accuracy with the residual error from the first game.  
Powerfully, the best guess we find from our second game will help us improve our answers to the first game, because the second game is basically [subtracting out](https://en.wikipedia.org/wiki/Gradient_boosting#Informal_introduction) the errors from the first game, and so can be added back in to increase predictive accuracy.  
And we can continue to do this for a third game, and a fourth game, and a fifth game, and so on until we've gotten every last bit of predictive accuracy out of our guesses.  
This doesn't mean we'll necessarily make perfect predictions on new students in the future, but it does mean that we won't leave any accuracy behind.

<h4>A Step-by-Step Recap of the Guessing Game</h4>
<ul>
<li>Define a Loss Function (this is the squared difference between prediction and reality, summed together for each student).
<li>Our starting game is the entire student data set.
</ul>
<ol>
<li>For each game (via GradientBoostingClassifier):
<ol>
<li>Define a hypothesis (this is our "guess").
<li>Update the hypothesis until it's as good as possible (via Gradient Descent).
<ol>
<li>Calculate the negative gradient of the loss function defined above (this indicates the "direction" we should change our guess towards).
<li>If the negative gradient isn't small enough, make a small change to the parameters of the hypothesis in the direction calculated above.
<li>If the negative gradient is small enough, we've reached the best guess for this game, and we're done with Gradient Descent.
</ol>
<li>Save this hypothesis, it's a good one, and we may "boost" it through subsequent iterations of games.
</ol>
<li>Subtract reality from our predictions for each student; this is our residual, and it will form the start of the next game.
<li>If the previous best-hypothesis helped our predictions, "boost" it.
<ol><li>Using the "residual" calculated above as the data, return to 1. above, and play another game.</ol>
<li>If the previous best-hypothesis did not help our predictions, we're done boosting, and done with the GradientBoostingClassifier algorithm.
</ol>
<ul><li>Return the best hypothesis, along with the "boosted" hypotheses we calculated from its subsequent residuals.  This is our finished GradientBoostingClassifier, and will be used to predict whether or not a student needs intervention in order to pass.</ul>

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [16]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV, RandomizedSearchCV
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {
    "learning_rate": np.arange(0.01, 0.02, 0.001)
    ,"n_estimators": range(60, 100, 1)
    ,"max_depth": range(30, 60, 1)
    ,"min_samples_leaf": range(20, 100, 1)
    #,"max_features": np.arange(0.6, 1.0, 0.01)
    #,"min_samples_split": range(1, 250, 1)
    ,"loss":["deviance","exponential"]
    #,"min_weight_fraction_leaf": np.arange(0.0, 0.2, 0.01)
    #,"subsample": np.arange(0.97, 0.85, -0.0033)
    #,"max_leaf_nodes": range(2,1000,10) 
}

# TODO: Initialize the classifier
clf = GradientBoostingClassifier(random_state = 42)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label="yes")

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = RandomizedSearchCV(clf, parameters, n_iter=100, scoring=f1_scorer, verbose=1, n_jobs=1, cv=10)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

# Report the parameters of the best estimator
print "Tuned model has the following parameters: " + str(grid_obj.best_params_)
print "Tuned model has the following cross-validated F1 score: " + str(grid_obj.best_score_)

Fitting 10 folds for each of 100 candidates, totalling 1000 fits


[Parallel(n_jobs=2)]: Done 104 tasks      | elapsed:    3.6s
[Parallel(n_jobs=2)]: Done 704 tasks      | elapsed:   24.3s
[Parallel(n_jobs=2)]: Done 1000 out of 1000 | elapsed:   33.7s finished


Made predictions in 0.0018 seconds.
Tuned model has a training F1 score of 0.8621.
Made predictions in 0.0007 seconds.
Tuned model has a testing F1 score of 0.7945.
Tuned model has the following parameters: {'n_estimators': 69, 'loss': 'exponential', 'learning_rate': 0.016999999999999994, 'max_depth': 44, 'min_samples_leaf': 24}
Tuned model has the following cross-validated F1 score: 0.834167559335


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

The final model's F<sub>1</sub> score for training was 0.8621.  
The final model's F<sub>1</sub> score for testing was 0.7945.

(see the below cell for the output of a variety of randomized trials with different parameters, each of which had approximately the same performance)

Actually, this is <font color="darkred">slightly worse</font> than the testing score for the best untuned model (KNeighbors, whose F<sub>1,testing</sub> score was 0.8138).  
The performance on the training set was worse, but that's not really the point of the exercise, although it does set a "ceiling" of sorts on the possible additional performance that can be gained.

It appears, then, that for this classification exercise, **KNeighbors** may be better suited than is the GradientBoostingClassifier (see subsequent cells for further proof of this assertion).

In a pure grid search scenario, it's unclear whether the final model is stuck in a local minimum of performance, in which case there are other pockets of the parameter-space that would lead to better performance on the testing set.  
However, by employing a RandomizedSearchCV, and repeating the trial numerous times, the search has a much better chance of locating high-performing local (or even global) minimums of performance.  
Hence, here I'm fairly confident that there is no set of parameters for the GradientBoostingClassifier which would yield better performance.




<h3>GradientBoostingClassifier Trial Results</h3>
Made predictions in 0.0015 seconds.  
Tuned model has a training F1 score of 0.8494.  
Made predictions in 0.0011 seconds.  
Tuned model has a **testing F1 score** of 0.7770.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.78000000000000003, 'learning_rate': 0.017000000000000001, 'min_samples_leaf': 22, 'n_estimators': 100, 'subsample': 0.79999999999999982, 'min_samples_split': 64, 'max_depth': 19}  
Tuned model has the following cross-validated F1 score: 0.828528408451  
  
Made predictions in 0.0057 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0029 seconds.  
Tuned model has a **testing F1 score** of 0.8028.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.5, 'learning_rate': 0.0040000000000000001, 'min_samples_leaf': 22, 'n_estimators': 940, 'subsample': 0.39999999999999947, 'min_samples_split': 34, 'max_depth': 2}  
Tuned model has the following cross-validated F1 score: 0.808515924705  
  
Made predictions in 0.0035 seconds.  
Tuned model has a training F1 score of 0.8657.  
Made predictions in 0.0022 seconds.  
Tuned model has a **testing F1 score** of 0.7770.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.78999999999999992, 'learning_rate': 0.012, 'min_samples_leaf': 33, 'n_estimators': 360, 'subsample': 0.83999999999999986, 'min_samples_split': 29, 'max_depth': 16}  
Tuned model has the following cross-validated F1 score: 0.816649583558  
  
Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8404.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7914.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.27999999999999997, 'learning_rate': 0.13900000000000001, 'min_samples_leaf': 3, 'n_estimators': 80, 'subsample': 0.26999999999999935, 'min_samples_split': 65, 'max_depth': 8}  
Tuned model has the following cross-validated F1 score: 0.816718020542  
  
Made predictions in 0.0011 seconds.  
Tuned model has a training F1 score of 0.8299.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7626.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.73999999999999988, 'learning_rate': 0.023, 'min_samples_leaf': 22, 'n_estimators': 80, 'subsample': 0.56999999999999962, 'min_samples_split': 79, 'max_depth': 33}  
Tuned model has the following cross-validated F1 score: 0.828836908933  
  
Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8833.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7660.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.77999999999999992, 'learning_rate': 0.14100000000000001, 'min_samples_leaf': 3, 'n_estimators': 60, 'subsample': 0.4399999999999995, 'min_samples_split': 82, 'max_depth': 12}  
Tuned model has the following cross-validated F1 score: 0.813103986818  
  
Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.8000.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.70999999999999996, 'learning_rate': 0.023, 'min_samples_leaf': 15, 'n_estimators': 45, 'subsample': 0.95999999999999996, 'min_samples_split': 82, 'max_depth': 38}  
Tuned model has the following cross-validated F1 score: 0.822982843027  
  
Made predictions in 0.0013 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7919.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.38999999999999996, 'learning_rate': 0.012, 'min_samples_leaf': 4, 'n_estimators': 80, 'subsample': 0.63999999999999968, 'min_samples_split': 63, 'max_depth': 8}  
Tuned model has the following cross-validated F1 score: 0.828817284885  
  
Made predictions in 0.0011 seconds.  
Tuned model has a training F1 score of 0.8246.  
Made predictions in 0.0008 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.67999999999999994, 'learning_rate': 0.018000000000000002, 'min_samples_leaf': 25, 'n_estimators': 75, 'subsample': 0.70999999999999974, 'min_samples_split': 90, 'max_depth': 22}  
Tuned model has the following cross-validated F1 score: 0.825569476809  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.8000.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.54999999999999993, 'learning_rate': 0.012, 'min_samples_leaf': 2, 'n_estimators': 80, 'subsample': 0.91999999999999993, 'min_samples_split': 90, 'max_depth': 29}
Tuned model has the following cross-validated F1 score: 0.825525333307  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8354.  
Made predictions in 0.0012 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.35999999999999999, 'learning_rate': 0.012, 'min_samples_leaf': 3, 'n_estimators': 85, 'subsample': 0.93999999999999995, 'min_samples_split': 110, 'max_depth': 33}  
Tuned model has the following cross-validated F1 score: 0.821714068606  
  
Made predictions in 0.0013 seconds.  
Tuned model has a training F1 score of 0.8390.  
Made predictions in 0.0007 seconds.  
Tuned model has a **testing F1 score** of 0.8000.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.72999999999999987, 'learning_rate': 0.0090000000000000011, 'min_samples_leaf': 7, 'n_estimators': 90, 'subsample': 0.94999999999999996, 'min_samples_split': 85, 'max_depth': 22}  
Tuned model has the following cross-validated F1 score: 0.829450745975  
  
Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8326.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7626.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.66999999999999993, 'learning_rate': 0.022000000000000002, 'min_samples_leaf': 29, 'n_estimators': 60, 'subsample': 0.90999999999999992, 'min_samples_split': 119, 'max_depth': 48}  
Tuned model has the following cross-validated F1 score: 0.829566487891  
  
Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8319.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.64999999999999991, 'learning_rate': 0.018000000000000002, 'min_samples_leaf': 3, 'n_estimators': 40, 'subsample': 0.74999999999999978, 'min_samples_split': 97, 'max_depth': 28}  
Tuned model has the following cross-validated F1 score: 0.833658437641  

Made predictions in 0.0011 seconds.  
Tuned model has a training F1 score of 0.8372.  
Made predictions in 0.0013 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.6399999999999999, 'learning_rate': 0.014999999999999999, 'min_samples_leaf': 7, 'n_estimators': 70, 'subsample': 0.70999999999999974, 'min_samples_split': 107, 'max_depth': 40}  
Tuned model has the following cross-validated F1 score: 0.835877410169  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8443.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.8000.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.88000000000000045, 'learning_rate': 0.014000000000000002, 'min_samples_leaf': 3, 'n_estimators': 75, 'subsample': 0.63999999999999968, 'min_samples_split': 152, 'max_depth': 26}  
Tuned model has the following cross-validated F1 score: 0.835877410169  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8430.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.8000.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.78000000000000036, 'learning_rate': 0.02, 'min_samples_leaf': 7, 'n_estimators': 70, 'subsample': 0.84999999999999987, 'min_samples_split': 172, 'max_depth': 34}  
Tuned model has the following cross-validated F1 score: 0.836422914422  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8435.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7945.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_features': 0.98000000000000032, 'learning_rate': 0.017000000000000001, 'min_samples_leaf': 14, 'n_estimators': 80, 'subsample': 0.75999999999999979, 'min_samples_split': 132, 'max_depth': 20}  
Tuned model has the following cross-validated F1 score: 0.834673181398  

Made predictions in 0.0017 seconds.  
Tuned model has a training F1 score of 0.8319.  
Made predictions in 0.0008 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'max_features': 0.94400000000000028, 'learning_rate': 0.0045000000000000005, 'min_samples_leaf': 3, 'n_estimators': 120, 'subsample': 0.7769999999999998, 'min_samples_split': 66, 'max_depth': 43}  
Tuned model has the following cross-validated F1 score: 0.832254221763  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8430.  
Made predictions in 0.0010 seconds.  
Tuned model has a **testing F1 score** of 0.7838.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.022799999999999924, 'min_samples_leaf': 9, 'n_estimators': 55, 'subsample': 0.86999999999999988, 'min_weight_fraction_leaf': 0.070000000000000007, 'max_features': 0.85000000000000042, 'min_samples_split': 140, 'max_depth': 25}  
Tuned model has the following cross-validated F1 score: 0.833882667564  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8383.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.021599999999999932, 'min_samples_leaf': 4, 'n_estimators': 51, 'subsample': 1.0, 'min_weight_fraction_leaf': 0.0060000000000000001, 'max_features': 0.88000000000000045, 'min_samples_split': 203, 'max_depth': 10}  
Tuned model has the following cross-validated F1 score: 0.832001853676  

Made predictions in 0.0011 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.012799999999999983, 'min_samples_leaf': 9, 'n_estimators': 87, 'subsample': 0.97999999999999998, 'min_weight_fraction_leaf': 0.019, 'max_features': 0.9400000000000005, 'min_samples_split': 189, 'max_depth': 16}  
Tuned model has the following cross-validated F1 score: 0.838871894014  

Made predictions in 0.0016 seconds.  
Tuned model has a training F1 score of 0.8849.  
Made predictions in 0.0008 seconds.  
Tuned model has a **testing F1 score** of 0.7692.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.021499999999999929, 'min_samples_leaf': 4, 'n_estimators': 73, 'subsample': 0.61999999999999966, 'min_weight_fraction_leaf': 0.070000000000000007, 'max_features': 0.8200000000000004, 'min_samples_split': 8, 'max_depth': 13}  
Tuned model has the following cross-validated F1 score: 0.83090697268  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8394.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7919.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.022399999999999927, 'min_samples_leaf': 17, 'n_estimators': 67, 'subsample': 0.85999999999999988, 'min_weight_fraction_leaf': 0.067000000000000004, 'max_features': 0.46000000000000008, 'min_samples_split': 174, 'max_depth': 25}  
Tuned model has the following cross-validated F1 score: 0.836170605769  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8383.  
Made predictions in 0.0009 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.015499999999999967, 'min_samples_leaf': 4, 'n_estimators': 71, 'subsample': 0.94999999999999996, 'min_weight_fraction_leaf': 0.0, 'max_features': 0.85000000000000042, 'min_samples_split': 224, 'max_depth': 16}  
Tuned model has the following cross-validated F1 score: 0.833365332874  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8419.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_leaf_nodes': 567, 'learning_rate': 0.015799999999999967, 'min_samples_leaf': 5, 'n_estimators': 91, 'subsample': 0.69999999999999973, 'min_weight_fraction_leaf': 0.0, 'max_features': 0.85000000000000042, 'min_samples_split': 185}  
Tuned model has the following cross-validated F1 score: 0.833428430577  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8383.  
Made predictions in 0.0011 seconds.  
Tuned model has a **testing F1 score** of 0.7947.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_leaf_nodes': 192, 'learning_rate': 0.012499999999999985, 'min_samples_leaf': 2, 'n_estimators': 103, 'subsample': 0.95999999999999996, 'min_weight_fraction_leaf': 0.050000000000000003, 'max_features': 0.64000000000000024, 'min_samples_split': 233}  
Tuned model has the following cross-validated F1 score: 0.833365332874  

Made predictions in 0.0013 seconds.  
Tuned model has a training F1 score of 0.8472.  
Made predictions in 0.0012 seconds.  
Tuned model has a **testing F1 score** of 0.7891.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_leaf_nodes': 82, 'learning_rate': 0.015299999999999968, 'min_samples_leaf': 10, 'n_estimators': 95, 'subsample': 0.5499999999999996, 'min_weight_fraction_leaf': 0.0, 'max_features': 0.91000000000000048, 'min_samples_split': 47}  
Tuned model has the following cross-validated F1 score: 0.832784082483  

Made predictions in 0.0011 seconds.  
Tuned model has a training F1 score of 0.8428.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7808.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_leaf_nodes': 692, 'learning_rate': 0.016799999999999961, 'min_samples_leaf': 7, 'n_estimators': 87, 'subsample': 0.93999999999999995, 'min_weight_fraction_leaf': 0.070000000000000007, 'max_features': 0.88000000000000045, 'min_samples_split': 149}  
Tuned model has the following cross-validated F1 score: 0.836796032346  

Made predictions in 0.0014 seconds.  
Tuned model has a training F1 score of 0.8522.  
Made predictions in 0.0016 seconds.  
Tuned model has a **testing F1 score** of 0.7808.  
Tuned model has the following parameters: {'loss': 'exponential', 'max_leaf_nodes': 272, 'learning_rate': 0.010999999999999999, 'min_samples_leaf': 8, 'n_estimators': 100, 'subsample': 0.96999999999999997, 'min_weight_fraction_leaf': 0.02, 'max_features': 0.70000000000000029, 'min_samples_split': 54}  
Tuned model has the following cross-validated F1 score: 0.837583519622  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8465.  
Made predictions in 0.0011 seconds.  
Tuned model has a **testing F1 score** of 0.7891.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.036999999999999977, 'min_samples_leaf': 4, 'n_estimators': 65, 'subsample': 0.96999999999999997, 'min_weight_fraction_leaf': 0.070000000000000007, 'max_features': 0.61000000000000021, 'min_samples_split': 149}  
Tuned model has the following cross-validated F1 score: 0.834382828721  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8596.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7891.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.01999999999999999, 'min_samples_leaf': 5, 'n_estimators': 65, 'subsample': 0.93999999999999995, 'min_weight_fraction_leaf': 0.02, 'max_features': 0.52000000000000013}  
Tuned model has the following cross-validated F1 score: 0.832072288098  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8534.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7891.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.015999999999999993, 'min_samples_leaf': 1, 'n_estimators': 70, 'subsample': 0.75999999999999979, 'min_weight_fraction_leaf': 0.059999999999999998, 'max_features': 0.8200000000000004}  
Tuned model has the following cross-validated F1 score: 0.832042882216  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8720.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7838.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.025999999999999988, 'min_samples_leaf': 5, 'n_estimators': 60, 'subsample': 0.91999999999999993, 'max_features': 0.91000000000000048}  
Tuned model has the following cross-validated F1 score: 0.834492081418  

Made predictions in 0.0009 seconds.  
Tuned model has a training F1 score of 0.8615.  
Made predictions in 0.0005 seconds.  
Tuned model has a **testing F1 score** of 0.7838.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.025999999999999988, 'min_samples_leaf': 9, 'n_estimators': 50, 'subsample': 0.91999999999999993, 'max_features': 0.59999999999999998}  
Tuned model has the following cross-validated F1 score: 0.833350074267  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8432.  
Made predictions in 0.0012 seconds.  
Tuned model has a **testing F1 score** of 0.8026.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.013999999999999997, 'min_samples_leaf': 19, 'n_estimators': 55, 'subsample': 0.91999999999999993, 'max_features': 0.73000000000000009, 'max_depth': 30}  
Tuned model has the following cross-validated F1 score: 0.831278179172  
  
Made predictions in 0.0013 seconds.  
Tuned model has a training F1 score of 0.8516.  
Made predictions in 0.0013 seconds.  
Tuned model has a **testing F1 score** of 0.7838.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.011999999999999999, 'min_samples_leaf': 17, 'n_estimators': 69, 'subsample': 0.87999999999999989, 'max_features': 0.88000000000000023, 'max_depth': 58}  
Tuned model has the following cross-validated F1 score: 0.832391580549  

Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8371.  
Made predictions in 0.0006 seconds.  
Tuned model has a **testing F1 score** of 0.7801.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.032999999999999981, 'min_samples_leaf': 50, 'n_estimators': 72, 'subsample': 0.93999999999999995, 'max_features': 0.65000000000000002, 'max_depth': 3}  
Tuned model has the following cross-validated F1 score: 0.832838429338  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8273.  
Made predictions in 0.0013 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.020999999999999991, 'min_samples_leaf': 43, 'n_estimators': 83, 'subsample': 0.92999999999999994, 'max_features': 0.84000000000000019, 'max_depth': 13}  
Tuned model has the following cross-validated F1 score: 0.831585549197  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8444.  
Made predictions in 0.0013 seconds.  
Tuned model has a **testing F1 score** of 0.7606.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.01999999999999999, 'min_samples_leaf': 26, 'n_estimators': 65, 'subsample': 0.96999999999999997, 'max_features': 0.63, 'max_depth': 46}  
Tuned model has the following cross-validated F1 score: 0.829212598335  
  
Made predictions in 0.0010 seconds.  
Tuned model has a training F1 score of 0.8318.  
Made predictions in 0.0012 seconds.  
Tuned model has a **testing F1 score** of 0.7770.  
Tuned model has the following parameters: {'loss': 'exponential', 'learning_rate': 0.030999999999999986, 'min_samples_leaf': 49, 'n_estimators': 74, 'subsample': 0.90070000000000061, 'max_features': 0.89000000000000024, 'max_depth': 38}  
Tuned model has the following cross-validated F1 score: 0.835407181645  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8246.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.014999999999999996, 'min_samples_leaf': 37, 'n_estimators': 77, 'subsample': 0.95680000000000009, 'max_features': 0.77000000000000013, 'max_depth': 46}  
Tuned model has the following cross-validated F1 score: 0.831035189431  
  
Made predictions in 0.0013 seconds.  
Tuned model has a training F1 score of 0.8273.  
Made predictions in 0.0014 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.015999999999999993, 'min_samples_leaf': 40, 'n_estimators': 93, 'subsample': 0.95350000000000013, 'max_features': 0.95000000000000029, 'max_depth': 40}  
Tuned model has the following cross-validated F1 score: 0.829073471903  
  
Made predictions in 0.0038 seconds.  
Tuned model has a training F1 score of 0.8717.  
Made predictions in 0.0018 seconds.  
Tuned model has a **testing F1 score** of 0.7724.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.014999999999999996, 'min_samples_leaf': 20, 'n_estimators': 90, 'subsample': 0.85120000000000107, 'max_features': 0.9600000000000003, 'max_depth': 46}  
Tuned model has the following cross-validated F1 score: 0.834536253109  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8246.  
Made predictions in 0.0007 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'loss': 'deviance', 'learning_rate': 0.015999999999999993, 'min_samples_leaf': 42, 'n_estimators': 74, 'subsample': 0.90730000000000055, 'max_features': 0.95000000000000029, 'max_depth': 56}  
Tuned model has the following cross-validated F1 score: 0.829073471903  
  
Made predictions in 0.0032 seconds.  
Tuned model has a training F1 score of 0.8547.
Made predictions in 0.0008 seconds.  
Tuned model has a **testing F1 score** of 0.7838.  
Tuned model has the following parameters: {'n_estimators': 97, 'loss': 'deviance', 'learning_rate': 0.010999999999999999, 'max_depth': 46, 'min_samples_leaf': 25}  
Tuned model has the following cross-validated F1 score: 0.834170280423  
  
Made predictions in 0.0012 seconds.  
Tuned model has a training F1 score of 0.8326.  
Made predictions in 0.0026 seconds.  
Tuned model has a **testing F1 score** of 0.7681.  
Tuned model has the following parameters: {'n_estimators': 93, 'loss': 'exponential', 'learning_rate': 0.017999999999999995, 'max_depth': 58, 'min_samples_leaf': 43}  
Tuned model has the following cross-validated F1 score: 0.828349957691  
  
Made predictions in 0.0018 seconds.  
Tuned model has a training F1 score of 0.8621.  
Made predictions in 0.0007 seconds.  
Tuned model has a **testing F1 score** of 0.7945.  
Tuned model has the following parameters: {'n_estimators': 69, 'loss': 'exponential', 'learning_rate': 0.016999999999999994, 'max_depth': 44, 'min_samples_leaf': 24}  
Tuned model has the following cross-validated F1 score: 0.834167559335

<h3>Alternative Model Tuning - KNeighborsClassifier</h3>
Additionally, I've ran the KNeighbors classification also with a RandomizedSearchCV, and it seems that F<sub>1</sub> for testing score is consistently above 0.8. (code and trial results follow).  This is a fairly strong indication that, for this classification task, KNeighbors is superior to the GradientBoostingClassifier discussed above.

In [21]:
# TODO: Create the parameters list you wish to tune
kn_parameters = {
    "n_neighbors": range(1,100,1)
    ,"algorithm":["kd_tree","ball_tree"]
    ,"leaf_size": range(1,100,1)
    ,"metric":["manhattan"]
    ,"weights":["distance"]
}

# TODO: Initialize the classifier
kn_clf = KNeighborsClassifier()

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label="yes")

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
kn_grid_obj = RandomizedSearchCV(kn_clf, kn_parameters, n_iter=100, scoring=f1_scorer, verbose=1, n_jobs=1, cv=10)

# TODO: Fit the grid search object to the training data and find the optimal parameters
kn_grid_obj = kn_grid_obj.fit(X_train, y_train)

# Get the estimator
kn_clf = kn_grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print("Tuned model has a training F1 score of {:.4f}.".format(predict_labels(kn_clf, X_train, y_train)))
print("Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(kn_clf, X_test, y_test)))

# Report the parameters of the best estimator
print("Tuned model has the following parameters: " + str(kn_grid_obj.best_params_))
print("Tuned model had cross-validated F1 score of: " + str(kn_grid_obj.best_score_))

Fitting 10 folds for each of 100 candidates, totalling 1000 fits
Made predictions in 0.0116 seconds.
Tuned model has a training F1 score of 1.0000.
Made predictions in 0.0048 seconds.
Tuned model has a testing F1 score of 0.8077.
Tuned model has the following parameters: {'n_neighbors': 21, 'metric': 'manhattan', 'weights': 'distance', 'leaf_size': 65, 'algorithm': 'ball_tree'}
Tuned model had cross-validated F1 score of: 0.813113482211


[Parallel(n_jobs=2)]: Done 1000 out of 1000 | elapsed:    5.3s finished


<h3>K-Nearest Neighbors Trial Results</h3>
<h4>cv = 10</h4>

Made predictions in 0.0206 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0073 seconds.  
Tuned model has a **testing F1 score** of 0.8025.  
Tuned model has the following parameters: {'leaf_size': 37, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 28, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.808156299841  
  
Made predictions in 0.1427 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0462 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 1, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.81092319151  
  
Made predictions in 0.0512 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0166 seconds.  
Tuned model has a **testing F1 score** of 0.8101.  
Tuned model has the following parameters: {'leaf_size': 3, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 40, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.809941706117  
  
Made predictions in 0.0219 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0070 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 14, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.812542239129  
  
Made predictions in 0.1442 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0456 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 1, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.81092319151  
  
Made predictions in 0.0502 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0168 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 3, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.81105363642  
  
Made predictions in 0.0142 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0048 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 40, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.809987812423  
  
Made predictions in 0.0155 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0062 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'leaf_size': 25, 'weights': 'distance', 'metric': 'manhattan', 'n_neighbors': 18, 'algorithm': 'kd_tree'}  
Tuned model had cross-validated F1 score of: 0.810851114041  
  
Made predictions in 0.0135 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0048 seconds.  
Tuned model has a **testing F1 score** of 0.8235.  
Tuned model has the following parameters: {'weights': 'distance', 'leaf_size': 61, 'algorithm': 'kd_tree', 'n_neighbors': 18, 'metric': 'manhattan'}  
Tuned model had cross-validated F1 score of: 0.809987812423  
  
<h4>cv=20</h4>
  
Made predictions in 0.0123 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0052 seconds.  
Tuned model has a **testing F1 score** of 0.8182.  
Tuned model has the following parameters: {'weights': 'distance', 'leaf_size': 91, 'algorithm': 'kd_tree', 'n_neighbors': 20, 'metric': 'manhattan'}  
Tuned model had cross-validated F1 score of: 0.814326355341  
  
Made predictions in 0.0161 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0062 seconds.  
Tuned model has a **testing F1 score** of 0.8289.  
Tuned model has the following parameters: {'weights': 'distance', 'algorithm': 'kd_tree', 'n_neighbors': 14, 'metric': 'manhattan'}  
Tuned model had cross-validated F1 score of: 0.813482589874  
  
Made predictions in 0.0164 seconds.  
Tuned model has a training F1 score of 1.0000.  
Made predictions in 0.0056 seconds.  
Tuned model has a **testing F1 score** of 0.8289.  
Tuned model has the following parameters: {'weights': 'distance', 'algorithm': 'kd_tree', 'n_neighbors': 14, 'metric': 'manhattan'}  
Tuned model had cross-validated F1 score of: 0.813482589874  

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.